cpp/arcticdb/processing/operation_types.hpp

```diff
@@ -373,18 +381,44 @@ bool operator()(int64_t t, uint64_t u) const {
 }
 };
 
-struct IsInOperator {
+struct MembershipOperator {
+protected:
+    /** Returns the high bits beyond what can be held in the signed integer type I, including I's sign bit. */
+    template<typename I>
+    static constexpr uint64_t incomparable_bits() {
+        return std::numeric_limits<uint64_t>::max() - static_cast<uint64_t>(std::numeric_limits<I>::max());
+    }
+
+    template<typename U>
+    static constexpr bool is_signed_int = std::is_integral_v<U> && std::is_signed_v<U>;
+
+public:
+    /** This is tighter than the signatures of the special handling operator()s below to reject argument types smaller
+     * than uint64 going down the special handling via type promotion. */
+    template<typename ColumnType, typename ValueSetBaseType>
+    static constexpr bool needs_uint64_special_handling =
+        (std::is_same_v<ColumnType, uint64_t> && is_signed_int<ValueSetBaseType>) ||
+        (std::is_same_v<ValueSetBaseType, uint64_t> && is_signed_int<ColumnType>);
+};
+
+/** Used as a dummy parameter to ensure we don't pick the non-special handling overloads by mistake. */
+struct UInt64SpecialHandlingTag {};
+
+struct IsInOperator: MembershipOperator {
 template<typename T, typename U>
 bool operator()(T t, const std::unordered_set<U>& u) const {
     return u.count(t) > 0;
 }
-bool operator()(uint64_t t, const std::unordered_set<int64_t>& u) const {
-    if (comparison::msb_set(t))
+
+template<typename U, typename=std::enable_if_t<is_signed_int<U>>>
+bool operator()(uint64_t t, const std::unordered_set<U>& u, UInt64SpecialHandlingTag = {}) const {
+    auto incomparable = t & incomparable_bits<U>();
```

Comment:
I don't think `incomparable_bits` conveys intent very well. I think this would be equivalent and more obvious:
`auto incomparable = t > static_cast<uint64_t>(std::numeric_limits<U>::max())`